api/s2n.h

```diff
@@ -596,6 +596,258 @@ extern int s2n_cert_get_utf8_string_from_extension_data_length(const uint8_t *ex
+ */
+S2N_API
+struct s2n_psk* s2n_external_psk_new();
+
+/**
+ * Frees the memory associated with the `s2n_psk` object.
+ *
+ * @param psk Pointer to the pre-shared key object to be freed.
+ */
+S2N_API 
+int s2n_psk_free(struct s2n_psk **psk);
+
+/**
+ * Sets the pre-shared key (PSK) identity for a given s2n_psk object.
+ * The pre-shared key identity is a unique identifier for the pre-shared secret.
+ * It is a non-secret value represented by a UTF-8 string.
+ *
+ * # Safety 
+ *
+ * The Pre-shared key identity is transmitted over the network unencrypted and is a non-secret value.
+ * Be sure to not include any confidential information as a part of the pre-shared key identity. 
+ * 
+ * Note that it is possible for multiple clients to use the same pre-shared key secret to connect to the server 
+ * in which case they would all share the pre-shared key identity associated with the shared pre-shared key secret.
+ *
+ * @param psk A pointer to pre-shared key object to be updated with the pre-shared key secret.
+ * @param identity The pre-shared key identity being set.
+ * @param identity_size The length of the pre-shared key identity being set.
+ */
+S2N_API 
+int s2n_psk_set_identity(struct s2n_psk *psk, const uint8_t *identity, uint16_t identity_size);
+
+/**
+ * Sets the out-of-band/externally provisioned pre-shared secret (PSK) for a given s2n_psk object.
+ * The pre-shared key secret is a secret value that is pre-shared between the server and client securely.
+ *
+ * # Safety
+ *
+ * When using an out-of-band provisioned pre-shared secret, a critical consideration is using 
+ * sufficient entropy during the pre-shared key generation. Deriving a shared secret from a password or 
+ * other low-entropy sources is not secure. A low-entropy secret, or password, is subject to 
+ * dictionary attacks based on the pre-shared key binder. Refer: RFC8446 Section-2.2 for more information.
+ *
+ * @param psk A pointer to s2n_psk object to be updated with the pre-shared secret.
+ * @param secret The pre-shared secret in raw bytes format.
+ * @param secret_size The length of the pre-shared secret being set.
+ */
+S2N_API 
+int s2n_psk_set_secret(struct s2n_psk *psk, const uint8_t *secret, uint16_t secret_size);
+
+/**
+ * Sets the pre-shared key (PSK) HMAC algorithm for a given s2n_psk object. 
+ * The supported pre-shared key HMAC algorithms are `SHA256` and `SHA384` 
+ * as listed in the enum `s2n_psk_hmac` above.
+ * 
+ * @param psk A pointer to the s2n_psk object to be updated with the pre-shared key HMAC algorithm.
+ * @param hmac The pre-shared key HMAC algorithm being set.  
+ */
+S2N_API 
+int s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac hmac);
+
+/**
+ * Appends a pre-shared key (PSK) object to the list of pre-shared keys supported by the s2n connection. 
+ * If a pre-shared key with a duplicate identity is found, an error is returned and the pre-shared key is not added to the list.
```

Comment:
I worry it might still be unclear what effect this actually does. Maybe mention how this PSK will be used by the client vs the server?

Reply:
This will be detailed and made clear in the Usage guide
```
Once the external PSK object is created and updated to include the PSK identity and secret,
the caller can add the external PSK to the connection using `s2n_connection_append_psk`
API. This API sets up the list of PSKs on the connection that will later be used by the 
pre-shared key extension.
```